src/wallet/rpcdump.cpp

```diff
@@ -806,266 +807,338 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+    const int64_t range_end = range.size() > 1 ? range[1].get_int64() : range_start;
+    if(range.size() > 2) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid range specified");
+    }
 
-                if (!pwallet->AddWatchOnly(script, timestamp)) {
-                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
-                }
+    FlatSigningProvider keys;
+    auto parsed_desc = Parse(descriptor, keys);
+    if (!parsed_desc) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Descriptor is invalid");
+    }
+    if (!parsed_desc->IsRange() && data.exists("range")) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Range should not be specified for an un-ranged descriptor");
+    } else if (parsed_desc->IsRange() && !data.exists("range")) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Descriptor is ranged, please specify the range");
+    }
+    FlatSigningProvider out_keys;
+    std::vector<CScript> script_pub_keys;
+    for (int i = range_start; i <= range_end; i++) {
+        std::vector<CScript> scripts_temp;
+        parsed_desc->Expand(i, keys, scripts_temp, out_keys);
+        script_pub_keys.insert(script_pub_keys.end(), scripts_temp.begin(), scripts_temp.end());
+    }
+    pwallet->MarkDirty();
+    // Import all scriptPubKeys
+    for (const CScript& script : script_pub_keys) {
+        if (::IsMine(*pwallet, script) == ISMINE_SPENDABLE) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
+        }
+        if (!pwallet->AddWatchOnly(script, timestamp)) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+        }
+        // add to address book or update label
+        CTxDestination destination;
+        if (ExtractDestination(script, destination) && IsValidDestination(destination)) {
+            pwallet->SetAddressBook(destination, label, "receive");
+        }
+    }
+    // Import all scripts from descriptors
+    for (auto const& x : out_keys.scripts) {
+        if (!pwallet->AddWatchOnly(x.second, timestamp)) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "Error adding script to wallet");
+        }
+        if (!pwallet->HaveCScript(x.first) && !pwallet->AddCScript(x.second)) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "Error adding script to wallet");
+        }
+    }
+    // Import all public keys from descriptors
+    for (auto const& x : out_keys.pubkeys) {
+        // This is necessary to force the wallet to import the pubKey
+        CScript raw_pubkey_script = GetScriptForRawPubKey(x.second);
+        if (::IsMine(*pwallet, raw_pubkey_script) == ISMINE_SPENDABLE) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this public key");
+        }
+        if (!pwallet->AddWatchOnly(raw_pubkey_script, timestamp)) {
```

Comment:
For the same reason as pointed out in the other PR, you can't do this; you're going to mark payments to individual multisig pubkeys as incoming payments.

Reply:
Restricting it from multisig(to avoid being tricked as you mention) would make this even more confusing to a user. 

Unfortunate.

Reply:
Mentioned on IRC, but what if we just never import public keys at all, and only either A) add the scriotPubKey as watch only or B) import the private key. If we have the private key then IMO it's less scary, because it's still "ours" and we can access the funds

Reply:
You're right; the same concern doesn't exist for private keys as you're obviously able to spend those coins anyway.

Reply:
Would not importing the pubkeys affect the wallet's ability to construct PSBT inputs?

Reply:
Only for things that have PKH/WPKH construction in them (GetPubKey is needed for those, which finds a pubkey based on pubkeyhash).